File: mirror-flower/graph.py
import matplotlib.pyplot as plt
import numpy as np
from enum import Enum

# Size of inputs vs. Delta Time plot

class Implementation(Enum):
    DOUBLY_LINKED_LIST = 0
    DYNAMIC_ARRAY = 1
    SKIP_LIST = 2
    TREE_SEQUENCE = 3

# Identify which implementation is being tested.

fileImple = open("../mirror-flower/outputs/implementation.txt", "r")
currImple = [str(line.strip()) for line in fileImple][0] 

complexitiesPerOperation ={
    "make" : ["O(l) worst-case", "O(l) worst-case", "O(l) expected", "O(l) worst-case"],
    "push_left" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(log(n)) worst-case"],
    "push_right" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(log(n)) worst-case"],
    "pop_left" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(log(n)) worst-case"],
    "pop_right" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(log(n)) worst-case"],
    "peek_left" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(1) worst-case"],
    "peek_right" : ["O(1) worst-case", "O(1) amortized", "O(log(n)) expected", "O(1) worst-case"],
    "size" : ["O(1) worst-case", "O(1) amortized", "O(log(n)) expected", "O(1) worst-case"],
    "empty" : ["O(1) worst-case", "O(1) amortized", "O(log(n)) expected", "O(1) worst-case"],
    "get" : ["O(1) worst-case", "O(1) amortized", "O(log(n)) expected", "O(log(n)) worst-case"],
    "set" : ["O(1) worst-case", "O(1) amortized", "O(log(n)) expected", "O(log(n)) worst-case"],
    "reverse" : ["O(1) worst-case", "O(1) amortized", "O(1) expected", "O(1) worst-case"]
    }

# Set the status of the grapher

match (currImple):
    case "DOUBLY_LINKED_LIST":
        status = Implementation.DOUBLY_LINKED_LIST
    case "DYNAMIC_ARRAY":
        status = Implementation.DYNAMIC_ARRAY
    case "SKIP_LIST":
        status = Implementation.SKIP_LIST
    case "TREE_SEQUENCE":
        status = Implementation.TREE_SEQUENCE

total_size = []
total_deltaTime = []

def plot_data():
    for operation in complexitiesPerOperation.keys():
        match operation:
            case "make":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/make.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "push_left":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/push_l.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "push_right":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/push_r.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "pop_left":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/pop_l.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "pop_right":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/pop_r.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "peek_left":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/peek_l.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "peek_right":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/peek_r.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "set":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/set.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)    
            case "get":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/get.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)         
            case "reverse":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/reverse.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)
                total_deltaTime.append(deltaTime)
                total_size.append(size)       
            case "empty":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/empty.txt", "r")
                for line in file:
                    extract(line, size, deltaTime)  
                total_deltaTime.append(deltaTime)
                total_size.append(size)     
            case "size":
                size = []
                deltaTime = []
                file = open("../mirror-flower/outputs/size.txt", "r")
                for line in file:
                    extract(line, size, deltaTime) 
                total_deltaTime.append(deltaTime)
                total_size.append(size)  
    return

def extract(line, size, deltaTime):
    parts = line.split("|")
    size.append(float(parts[0].strip()))
    deltaTime.append(float(parts[1].strip()))
    return size, deltaTime

def graph_settings():
    fig, axes1 =  plt.subplots(2,3)
    axes1[0, 0].plot(total_size[0],total_deltaTime[0],label=complexitiesPerOperation["make"][status.value])
    axes1[0, 0].set_title("Operation: make")
    axes1[0, 0].grid()
    axes1[0, 0].legend()

    axes1[0, 1].plot(total_size[1],total_deltaTime[1],label=complexitiesPerOperation["push_left"][status.value])
    axes1[0, 1].set_title("Operation: push_left")
    axes1[0, 1].grid()
    axes1[0, 1].legend()

    axes1[0, 2].plot(total_size[2],total_deltaTime[2],label=complexitiesPerOperation["push_right"][status.value])
    axes1[0, 2].set_title("Operation: push_right")
    axes1[0, 2].grid()
    axes1[0, 2].legend()

    axes1[1, 1].plot(total_size[3],total_deltaTime[3],label=complexitiesPerOperation["pop_left"][status.value])
    axes1[1, 1].set_title("Operation: pop_left")
    axes1[1, 1].grid()
    axes1[1, 1].legend()

    axes1[1, 2].plot(total_size[4],total_deltaTime[4],label=complexitiesPerOperation["pop_right"][status.value])
    axes1[1, 2].set_title("Operation: pop_right")
    axes1[1, 2].grid()
    axes1[1, 2].legend()

    axes1[1, 0].plot(total_size[5],total_deltaTime[5],label=complexitiesPerOperation["reverse"][status.value])
    axes1[1, 0].set_title("Operation: reverse")
    axes1[1, 0].grid()
    axes1[1, 0].legend()

    
    fig, axes2 =  plt.subplots(2, 3)
    axes2[0, 0].plot(total_size[6],total_deltaTime[6],label=complexitiesPerOperation["peek_left"][status.value])
    axes2[0, 0].set_title("Operation: peek_left")
    axes2[0, 0].grid()
    axes2[0, 0].legend()

    axes2[0, 1].plot(total_size[7],total_deltaTime[7],label=complexitiesPerOperation["peek_right"][status.value])
    axes2[0, 1].set_title("Operation: peek_right")
    axes2[0, 1].grid()
    axes2[0, 1].legend()

    axes2[0, 2].plot(total_size[8],total_deltaTime[8],label=complexitiesPerOperation["size"][status.value])
    axes2[0, 2].set_title("Operation: size")
    axes2[0, 2].grid()
    axes2[0, 2].legend()

    axes2[1, 0].plot(total_size[9],total_deltaTime[9],label=complexitiesPerOperation["get"][status.value])
    axes2[1, 0].set_title("Operation: get")
    axes2[1, 0].grid()
    axes2[1, 0].legend()

    axes2[1, 1].plot(total_size[10],total_deltaTime[10],label=complexitiesPerOperation["set"][status.value])
    axes2[1, 1].set_title("Operation: set")
    axes2[1, 1].grid()
    axes2[1, 1].legend()

    axes2[1, 2].plot(total_size[11],total_deltaTime[11],label=complexitiesPerOperation["empty"][status.value])
    axes2[1, 2].set_title("Operation: empty")
    axes2[1, 2].grid()
    axes2[1, 2].legend()
    return 

def main():
    plot_data()
    graph_settings()
    plt.show()
main()

File: mirror-flower/graph.c
#include <stdio.h>
#include <stdlib.h>
#include <float.h>
#include <time.h>
#include <windows.h>
#include <sys/timeb.h>
#include "../H_global.h"

// Implementation selection
#if IMPLEMENTATION == DOUBLY_LINKED_lIST
    #include "../doubly-linked-list/d-linked-list.c"
    typedef list Reflection;
#elif IMPLEMENTATION == DYNAMIC_ARRAY
    #include "../dynamic-array/dynamic-array.c"
    typedef dynamic_array Reflection;
#elif IMPLEMENTATION == SKIP_LIST
    #include "../skip-list/skip-list.c"
    typedef SkipList Reflection;
#elif IMPLEMENTATION == TREE_SEQUENCE
    #include "../tree-sequence/p_tree_list.c"
    typedef PTreeList Reflection;
#endif

// Timer definitions
typedef LARGE_INTEGER RECORDED_TIME;
typedef double PROCESSED_TIME;
uint64_t freq;

// Timer initialization
void _TIME_init() {
    LARGE_INTEGER f;
    QueryPerformanceFrequency(&f);
    freq = (uint64_t)((uint64_t)(f.HighPart) << 32) | (uint32_t)f.LowPart;
}

void _TIME(RECORDED_TIME* cRef) {
    QueryPerformanceCounter(cRef);
}

PROCESSED_TIME _PROCESSTIME(RECORDED_TIME a, RECORDED_TIME b) {
    uint64_t bQ = (uint64_t)((uint64_t)(b.HighPart) << 32) | (uint32_t)b.LowPart;
    uint64_t aQ = (uint64_t)((uint64_t)(a.HighPart) << 32) | (uint32_t)a.LowPart;
    uint64_t dt = (bQ - aQ);
    PROCESSED_TIME final = (double)(dt) / (double)freq;
    return final < 0.0L ? 0.0L : final;
}

// Data structure for writing results
typedef struct _WriteData {
    char* operation;
    LENGTH n;
    PROCESSED_TIME c;
} WriteData;

WriteData NEW_WRITE(char* operation, LENGTH n, PROCESSED_TIME c) {
    WriteData writeData = *((WriteData*)malloc(sizeof(WriteData)));
    writeData.operation = operation;
    writeData.n = n;
    writeData.c = c;
    return writeData;
}

void EXPORT_DELTA_TIME(FILE** output, WriteData wd, bool newLine) {
    FILE* target = NULL;
    
    if (strcmp(wd.operation, "make") == 0) target = output[0];
    else if (strcmp(wd.operation, "push_left") == 0) target = output[1];
    else if (strcmp(wd.operation, "push_right") == 0) target = output[2];
    else if (strcmp(wd.operation, "pop_left") == 0) target = output[3];
    else if (strcmp(wd.operation, "pop_right") == 0) target = output[4];
    else if (strcmp(wd.operation, "peek_left") == 0) target = output[5];
    else if (strcmp(wd.operation, "peek_right") == 0) target = output[6];
    else if (strcmp(wd.operation, "set") == 0) target = output[7];
    else if (strcmp(wd.operation, "get") == 0) target = output[8];
    else if (strcmp(wd.operation, "reverse") == 0) target = output[9];
    else if (strcmp(wd.operation, "empty") == 0) target = output[10];
    else if (strcmp(wd.operation, "size") == 0) target = output[11];

    if (target) {
        fprintf(target, "%zu | %.9lf", wd.n, wd.c);
        if (newLine) fprintf(target, "\n");
    }
}

// Operation execution and timing
void OPERATION(FILE** output, Reflection* list, char* operation) {
    RECORDED_TIME _rec, rec;
    LENGTH n = size(list);
    
    // Warmup iterations
    for (int i = 0; i < 100; i++) {
        if (strcmp(operation, "size") == 0) size(list);
        else if (strcmp(operation, "empty") == 0) empty(list);
        else if (strcmp(operation, "reverse") == 0) reverse(list);
        else if (strcmp(operation, "get") == 0 && n > 0) get(list, (LENGTH)rand() % n);
        else if (strcmp(operation, "set") == 0 && n > 0) set(list, (LENGTH)rand() % n, rand());
        else if (strcmp(operation, "peek_left") == 0 && n > 0) peek_left(list);
        else if (strcmp(operation, "peek_right") == 0 && n > 0) peek_right(list);
        else if (strcmp(operation, "push_left") == 0) push_left(list, rand());
        else if (strcmp(operation, "push_right") == 0) push_right(list, rand());
        else if (strcmp(operation, "pop_left") == 0 && n > 0) pop_left(list);
        else if (strcmp(operation, "pop_right") == 0 && n > 0) pop_right(list);
    }

    // Actual measurement
    _TIME(&_rec);
    if (strcmp(operation, "size") == 0) {
        size(list);
    } else if (strcmp(operation, "empty") == 0) {
        empty(list);
    } else if (strcmp(operation, "reverse") == 0) {
        reverse(list);
    } else if (strcmp(operation, "get") == 0) {
        if (n == 0) return;
        get(list, (LENGTH)rand() % n);
    } else if (strcmp(operation, "set") == 0) {
        if (n == 0) return;
        set(list, (LENGTH)rand() % n, rand());
    } else if (strcmp(operation, "peek_left") == 0) {
        if (n == 0) return;
        peek_left(list);
    } else if (strcmp(operation, "peek_right") == 0) {
        if (n == 0) return;
        peek_right(list);
    } else if (strcmp(operation, "push_left") == 0) {
        push_left(list, rand());
    } else if (strcmp(operation, "push_right") == 0) {
        push_right(list, rand());
    } else if (strcmp(operation, "pop_left") == 0) {
        if (n == 0) return;
        pop_left(list);
    } else if (strcmp(operation, "pop_right") == 0) {
        if (n == 0) return;
        pop_right(list);
    }
    _TIME(&rec);
    
    EXPORT_DELTA_TIME(output, NEW_WRITE(operation, n, _PROCESSTIME(_rec, rec)), true);
}

int main() {
    printf("<< Grapher >>\n");
    printf("> Initializing variables...\n");

    // Write implementation type
    FILE* implementation = fopen("../mirror-flower/outputs/implementation.txt", "w+");
    if (IMPLEMENTATION == DOUBLY_LINKED_LIST) fprintf(implementation, "DOUBLY_LINKED_LIST");
    else if (IMPLEMENTATION == DYNAMIC_ARRAY) fprintf(implementation, "DYNAMIC_ARRAY");
    else if (IMPLEMENTATION == SKIP_LIST) fprintf(implementation, "SKIP_LIST");
    else if (IMPLEMENTATION == TREE_SEQUENCE) fprintf(implementation, "TREE_SEQUENCE");
    fclose(implementation);

    // Initialize timer and random seed
    _TIME_init();
    srand(time(NULL));

    // Initialize output files
    FILE** output = malloc(12 * sizeof(FILE));
    output[0] = fopen("../mirror-flower/outputs/make.txt", "w+");
    output[1] = fopen("../mirror-flower/outputs/push_l.txt", "w+");
    output[2] = fopen("../mirror-flower/outputs/push_r.txt", "w+");
    output[3] = fopen("../mirror-flower/outputs/pop_l.txt", "w+");
    output[4] = fopen("../mirror-flower/outputs/pop_r.txt", "w+");
    output[5] = fopen("../mirror-flower/outputs/peek_l.txt", "w+");
    output[6] = fopen("../mirror-flower/outputs/peek_r.txt", "w+");
    output[7] = fopen("../mirror-flower/outputs/set.txt", "w+");
    output[8] = fopen("../mirror-flower/outputs/get.txt", "w+");
    output[9] = fopen("../mirror-flower/outputs/reverse.txt", "w+");
    output[10] = fopen("../mirror-flower/outputs/empty.txt", "w+");
    output[11] = fopen("../mirror-flower/outputs/size.txt", "w+");

    printf("> Done.\n");
    printf("> Generating plot points...\n");

    // Test creation with different sizes
    for (LENGTH n = 0; n < 50000; n += 100) {
        DATA* seq = (DATA*)malloc(n * sizeof(DATA));
        for (LENGTH i = 0; i < n; i++) {
            seq[i] = rand();
        }
        RECORDED_TIME _rec, rec;
        _TIME(&_rec);
        Reflection* list = make(n, seq);
        _TIME(&rec);
        EXPORT_DELTA_TIME(output, NEW_WRITE("make", n, _PROCESSTIME(_rec, rec)), true);
        free(seq);
        // Clean up list based on implementation
        // Add appropriate cleanup here
    }

    // Initialize empty list for operations testing
    Reflection* list = make(0, NULL);

    // Test operations with increasing sizes
    LENGTH upper = 20000;
    for (int i = 0; i < upper; i++) {
        char* operations[] = {
            "push_left", "push_right", "pop_left", "pop_right",
            "peek_left", "peek_right", "get", "set",
            "size", "empty", "reverse"
        };
        
        for (int j = 0; j < 11; j++) {
            OPERATION(output, list, operations[j]);
        }
    }

    // Cleanup
    for (int i = 0; i < 12; i++) {
        fclose(output[i]);
    }
    free(output);

    printf("> Done.\n");
    return 0;
}